lib/storage/src/rbac/mod.rs

```diff
@@ -1,9 +1,209 @@
-pub mod access;
-pub mod access_level;
-pub mod access_token;
-#[allow(dead_code)]
-pub mod collection_access;
-#[allow(dead_code)]
-pub mod error;
-
 mod ops_checks;
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+use segment::json_path::JsonPath;
+use segment::types::ValueVariants;
+use serde::{Deserialize, Serialize};
+
+use crate::content_manager::errors::StorageError;
+
+/// A structure that defines access rights.
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+#[serde(untagged)]
+pub enum Access {
+    /// Cluster-wide access.
+    Cluster(ClusterAccessMode),
+    /// Access to specific collections.
+    Collection(AccessCollection),
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct AccessCollection(pub Vec<AccessCollectionRule>);
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct AccessCollectionRule {
+    /// Collection names that are allowed to be accessed
+    pub collections: Vec<String>,
+
+    pub access: CollectionAccessMode,
+
+    /// Payload constraints.
+    /// An object where each key is a JSON path, and each value is JSON value.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub payload: Option<PayloadConstraint>,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum ClusterAccessMode {
+    /// Read-only access to the whole cluster.
+    #[serde(rename = "r")]
+    Read,
+
+    /// Read and write access to the whole cluster.
+    #[serde(rename = "rw")]
+    ReadWrite,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum CollectionAccessMode {
+    /// Read-only access to a collection.
+    #[serde(rename = "r")]
+    Read,
+
+    /// Read and write access to a collection, with some restrictions.
+    #[serde(rename = "rw")]
+    ReadWrite,
+
+    /// Read and write access to a collection, with no restrictions.
+    #[serde(rename = "m")]
+    Manage,
+}
```

Comment:
We can add collection `manage` access later. For now, and to allow this extensibility without (hopefully) a confusing interface, we might want to rename full cluster access to `Manage`.


```suggestion
pub enum ClusterAccessMode {
    /// Read-only access to the whole cluster.
    #[serde(rename = "r")]
    Read,

    /// Read and write access to the whole cluster.
    #[serde(rename = "m")]
    Manage,
}

#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
pub enum CollectionAccessMode {
    /// Read-only access to a collection.
    #[serde(rename = "r")]
    Read,

    /// Read and write access to a collection, with some restrictions.
    #[serde(rename = "rw")]
    ReadWrite,
}
```